### PR TITLE
[BugFix] Fix the bug of not locking when cleaning txn_meta

### LINKED ISSUE #13283: [Bug] Be crashes repeatedly and fails to start
### Steps to reproduce the behavior (Required)
1. CREATE TABLE '...'
2. INSERT INTO '....'
3. SELECT  '....'

### Expected behavior (Required)

run success
### Real behavior (Required)
be crash and can't start
image

### StarRocks version (Required)
 - 2.2.x，2.3.x，2.4.x
 

### Bug Analysis
The core stack is shown as below
image
BE crashes when do load partial rowset, and the crash reason is `partial_rowset_footer` in rowset_meta is empty.  And when we load a partial rowset, we will access to `partial_rowset_footer` which get a NPE.
image

However, the existence of txn_meta in rowset_meta but not partial_rowset_footer in txn_meta should not happen, and the partial rowset is clone from other be.
image

When we do apply of partial rowset, we will clear `txn_meta` at last, but this operation is not locked. So if the clone operation happens at the same time, the incomplete rowset_meta may be read which will cause the following error.
image
image

### PR DESCRIPTION
**Pull request by sevev**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes https://github.com/StarRocks/starrocks/issues/13283

## Problem Summary(Required) ：

When we do apply of partial rowset, we will clear txn_meta at last, but this operation is not locked. So if the clone operation happens at the same time, the incomplete rowset_meta may be read which will cause the following error.
```
// This function is unlock
Status RowsetUpdateState::_update_rowset_meta(Tablet* tablet, Rowset* rowset) {
    rowset->rowset_meta()->clear_txn_meta();
    auto& rowset_meta_pb = rowset->rowset_meta()->get_meta_pb();
    return TabletMetaManager::write_rowset_meta(tablet->data_dir(), tablet->tablet_id(), rowset_meta_pb, string());
}
```

```
Status SnapshotManager::make_snapshot_on_tablet_meta(SnapshotTypePB snapshot_type, const std::string& snapshot_dir,
                                                     const TabletSharedPtr& tablet,
                                                     const std::vector<RowsetMetaSharedPtr>& rowset_metas,
                                                     int64_t snapshot_version, int32_t snapshot_format) {
...
    // we will access rowset_meta in this function
    for (const auto& rowset_meta : rowset_metas) {
        RowsetMetaPB& meta_pb = snapshot_meta.rowset_metas().emplace_back();
        rowset_meta->to_rowset_pb(&meta_pb);
    }
...
```

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Comment by sevev** [comment]
I will modify this pr again, do not approve for now plz

**Comment by sevev** [comment]
@decster PTAL

File: be/src/storage/tablet_meta_manager.cpp [hunk 1]
```diff
@@ -804,7 +804,8 @@ Status TabletMetaManager::rowset_iterate(DataDir* store, TTabletId tablet_id, co
 Status TabletMetaManager::apply_rowset_commit(DataDir* store, TTabletId tablet_id, int64_t logid,
                                               const EditVersion& version,
                                               vector<std::pair<uint32_t, DelVectorPtr>>& delvecs,
-                                              const PersistentIndexMetaPB& index_meta, bool enable_persistent_index) {
+                                              const PersistentIndexMetaPB& index_meta, bool enable_persistent_index,
+                                              Rowset* rowset) {
```
Comment (decster) [inline be/src/storage/tablet_meta_manager.cpp:808]: passing a rowset ptr without meaning seems confusing, is the intention to persist rowset meta pb or `clear rowset txn meta then persist if rowset not empty`? the intention is just to complex. 
I think passing a rowset meta pb(or null) is more clear.

Comment (sevev) [reply be/src/storage/tablet_meta_manager.cpp]: > passing a rowset ptr without meaning seems confusing, is the intention to persist rowset meta pb or `clear rowset txn meta then persist if rowset not empty`? the intention is just to complex. I think passing a rowset meta pb(or null) is more clear.

done, changed in https://github.com/StarRocks/starrocks/pull/13285/commits/0c44e1b2804661258d1d9495f390201d4caa5f94

Comment (sevev) [reply be/src/storage/tablet_meta_manager.cpp]: run starrocks_be_unittest

**Comment by sevev** [comment]
run starrocks_be_unittest

**Review by decster — DISMISSED** [review]

**Review by chaoyli — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by sevev** [comment]
run pr-title-check

**Comment by wanpengfei-git** [comment]
@Mergify rebase

**Comment by sevev** [comment]
run all

**Comment by dulong41** [comment]
run starrocks_admit_test

**Review by decster — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Comment by sevev** [comment]
@mergify backport branch-2.4

**Comment by sevev** [comment]
@mergify backport branch-2.4 branch-2.3 branch-2.2 branch-2.5

**Pull request merged**